File: frameworks/AgentsMesh/packages/electron-adapter/src/repository.ts
import { invoke } from "./invoke";
import { coerceConnectResponse } from "./connect-response";
import { unwrapIpcServiceError } from "./connect-ipc-error";
import type { IRepositoryService, IRepoState } from "@agentsmesh/service-interface";
import { fromBinary, toBinary, create } from "@bufbuild/protobuf";
import {
  ReplaceCachedRepositoriesRequestSchema,
  SetCurrentRepoRequestSchema,
  ReplaceBranchesRequestSchema,
  InsertRepositoryRequestSchema,
  PatchRepositoryRequestSchema,
} from "@agentsmesh/proto/repo_state/v1/repo_state_pb";
import { ListRepositoriesResponseSchema } from "@agentsmesh/proto/repository/v1/repository_pb";
import { repositoryToCache, branchToCache } from "./projections/repository";
import { repositoriesBytes } from "./repository_cache_to_bytes";

// Web's wasm-side `WasmRepositoryService` exposes `<verb>Connect(bytes)`
// methods. RepositoryService in Rust core is a thin proxy (no cache, no
// business logic — see ADR 2026-05-24-service-binding-matrix.md "Decision
// 1: thin-proxy services don't require binding symmetry"). Desktop renderer
// reaches the backend through `connectCall` generic IPC proxy (main owns
// the auth header injection + URL prefix) — adding a dedicated napi binding
// would be form symmetry without functional gain, so we go directly to the
// backend Connect endpoint through main process.
async function connectCall(method: string, request: Uint8Array): Promise<Uint8Array> {
  try {
    const resp = await invoke<number[] | Uint8Array>(
      "connectCall",
      "proto.repository.v1.RepositoryService",
      method,
      Array.from(request),
    );
    return coerceConnectResponse(resp);
  } catch (e) {
    throw unwrapIpcServiceError(e);
  }
}

// Service-is-State-superset 模式（对齐 ElectronPodService）：内部 cache 由 Service 持有，
// provider 让 repoState 别名同一实例，renderer 端 useRepositories() 读到的就是这里的 cache。
export class ElectronRepositoryService implements IRepositoryService, IRepoState {
  private _repositoriesCache = "[]";
  private _currentRepoCache: string | null = null;
  private _branchesCache = "[]";

  // ===== IRepoState =====

  repositories_json(): string { return this._repositoriesCache; }
  current_repo_json(): unknown { return this._currentRepoCache; }
  branches_json(): string { return this._branchesCache; }

  // Read side (B, zero-JSON): re-encode renderer cache into state proto bytes.
  repositories_bytes(): Uint8Array { return repositoriesBytes(this._repositoriesCache); }

  // Fetch→state (B): wire Repository == cache Repository. Decode wire response →
  // renderer cache (sync), then fan the SAME repos to main via the existing
  // replace_cached_repositories NAPI (best-effort; main repo state is non-authoritative).
  apply_fetched_repositories(respBytes: Uint8Array): void {
    const resp = fromBinary(ListRepositoriesResponseSchema, respBytes);
    this._repositoriesCache = JSON.stringify(resp.items.map(repositoryToCache));
    const reqBytes = toBinary(ReplaceCachedRepositoriesRequestSchema,
      create(ReplaceCachedRepositoriesRequestSchema, { repositories: resp.items }));
    void invoke<void>("repoReplaceCachedRepositories", Array.from(reqBytes)).catch(() => undefined);
  }

  remove_repository(id: string): void {
    const repos = JSON.parse(this._repositoriesCache) as { id: number }[];
    this._repositoriesCache = JSON.stringify(repos.filter(r => String(r.id) !== id));
  }

  // Proto-bytes mutators — decode locally + update JS cache synchronously,
  // then fire-and-forget NAPI sync. Mirrors ElectronRunnerService.

  set_current_repo_proto(reqBytes: Uint8Array): void {
    const req = fromBinary(SetCurrentRepoRequestSchema, reqBytes);
    this._currentRepoCache = req.repository ? JSON.stringify(repositoryToCache(req.repository)) : null;
    void invoke<void>("repoSetCurrentRepoProto", Array.from(reqBytes)).catch(() => undefined);
  }

  replace_branches(reqBytes: Uint8Array): void {
    const req = fromBinary(ReplaceBranchesRequestSchema, reqBytes);
    this._branchesCache = JSON.stringify(req.branches.map(branchToCache));
    void invoke<void>("repoReplaceBranches", Array.from(reqBytes)).catch(() => undefined);
  }

  insert_repository(reqBytes: Uint8Array): void {
    const req = fromBinary(InsertRepositoryRequestSchema, reqBytes);
    if (req.repository) {
      const repos = JSON.parse(this._repositoriesCache) as unknown[];
      repos.push(repositoryToCache(req.repository));
      this._repositoriesCache = JSON.stringify(repos);
    }
    void invoke<void>("repoInsertRepository", Array.from(reqBytes)).catch(() => undefined);
  }

  patch_repository(reqBytes: Uint8Array): void {
    const req = fromBinary(PatchRepositoryRequestSchema, reqBytes);
    if (req.repository) {
      const patched = repositoryToCache(req.repository);
      const repos = JSON.parse(this._repositoriesCache) as { id: number }[];
      const idx = repos.findIndex(r => String(r.id) === req.id);
      if (idx >= 0) repos[idx] = patched as { id: number };
      this._repositoriesCache = JSON.stringify(repos);
    }
    void invoke<void>("repoPatchRepository", Array.from(reqBytes)).catch(() => undefined);
  }

  // ===== IRepositoryService =====

  async list(): Promise<string> {
    return invoke<string>("repositoryList");
  }

  async get(id: bigint): Promise<string> {
    return invoke<string>("repositoryGet", Number(id));
  }

  async create(json: string): Promise<string> {
    return invoke<string>("repositoryCreate", json);
  }

  async update(id: bigint, json: string): Promise<string> {
    return invoke<string>("repositoryUpdate", Number(id), json);
  }

  async delete(id: bigint): Promise<void> {
    await invoke<void>("repositoryDelete", Number(id));
  }

  async list_branches(id: bigint): Promise<string> {
    return invoke<string>("repositoryListBranches", Number(id));
  }

  async sync_branches(id: bigint, json: string): Promise<string> {
    return invoke<string>("repositorySyncBranches", Number(id), json);
  }

  async list_merge_requests(id: bigint, branch?: string | null, state?: string | null): Promise<string> {
    return invoke<string>("repositoryListMergeRequests", Number(id), branch, state);
  }

  async register_webhook(id: bigint): Promise<void> {
    await invoke<void>("repositoryRegisterWebhook", Number(id));
  }

  async delete_webhook(id: bigint): Promise<void> {
    await invoke<void>("repositoryDeleteWebhook", Number(id));
  }

  async get_webhook_secret(id: bigint): Promise<string> {
    return invoke<string>("repositoryGetWebhookSecret", Number(id));
  }

  async get_webhook_status(id: bigint): Promise<string> {
    return invoke<string>("repositoryGetWebhookStatus", Number(id));
  }

  async mark_webhook_configured(id: bigint): Promise<void> {
    await invoke<void>("repositoryMarkWebhookConfigured", Number(id));
  }

  // ── Connect-RPC binary surface (mirrors WasmRepositoryService) ──

  listRepositoriesConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("ListRepositories", request);
  }
  getRepositoryConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("GetRepository", request);
  }
  createRepositoryConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("CreateRepository", request);
  }
  updateRepositoryConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("UpdateRepository", request);
  }
  deleteRepositoryConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("DeleteRepository", request);
  }
  listRepositoryBranchesConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("ListRepositoryBranches", request);
  }
  syncRepositoryBranchesConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("SyncRepositoryBranches", request);
  }
  listRepositoryMergeRequestsConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("ListRepositoryMergeRequests", request);
  }
  registerRepositoryWebhookConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("RegisterRepositoryWebhook", request);
  }
  deleteRepositoryWebhookConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("DeleteRepositoryWebhook", request);
  }
  getRepositoryWebhookStatusConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("GetRepositoryWebhookStatus", request);
  }
  getRepositoryWebhookSecretConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("GetRepositoryWebhookSecret", request);
  }
  markRepositoryWebhookConfiguredConnect(request: Uint8Array): Promise<Uint8Array> {
    return connectCall("MarkRepositoryWebhookConfigured", request);
  }
}

File: frameworks/AgentsMesh/clients/core/crates/api-client/src/modules/repository.rs
use crate::ApiClient;
use crate::connect_call::connect_call;
use crate::error::ApiError;
use agentsmesh_types::proto_repository_v1 as repo_proto;

// =============================================================================
// Connect-RPC (binary wire). See proto-naming-conventions.md §2.5.
// =============================================================================
//
// These methods call the Connect handlers in
// backend/internal/api/connect/repository/. Procedure paths derive from
// `proto.repository.v1.RepositoryService.<Method>` (conventions §12).

impl ApiClient {
    pub async fn list_repositories_connect(
        &self,
        req: &repo_proto::ListRepositoriesRequest,
    ) -> Result<repo_proto::ListRepositoriesResponse, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/ListRepositories",
            req,
        )
        .await
    }

    pub async fn get_repository_connect(
        &self,
        req: &repo_proto::GetRepositoryRequest,
    ) -> Result<repo_proto::Repository, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/GetRepository",
            req,
        )
        .await
    }

    pub async fn create_repository_connect(
        &self,
        req: &repo_proto::CreateRepositoryRequest,
    ) -> Result<repo_proto::Repository, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/CreateRepository",
            req,
        )
        .await
    }

    pub async fn update_repository_connect(
        &self,
        req: &repo_proto::UpdateRepositoryRequest,
    ) -> Result<repo_proto::Repository, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/UpdateRepository",
            req,
        )
        .await
    }

    pub async fn delete_repository_connect(
        &self,
        req: &repo_proto::DeleteRepositoryRequest,
    ) -> Result<repo_proto::DeleteRepositoryResponse, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/DeleteRepository",
            req,
        )
        .await
    }

    pub async fn list_repository_branches_connect(
        &self,
        req: &repo_proto::ListRepositoryBranchesRequest,
    ) -> Result<repo_proto::ListRepositoryBranchesResponse, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/ListRepositoryBranches",
            req,
        )
        .await
    }

    pub async fn sync_repository_branches_connect(
        &self,
        req: &repo_proto::SyncRepositoryBranchesRequest,
    ) -> Result<repo_proto::ListRepositoryBranchesResponse, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/SyncRepositoryBranches",
            req,
        )
        .await
    }

    pub async fn list_repository_merge_requests_connect(
        &self,
        req: &repo_proto::ListRepositoryMergeRequestsRequest,
    ) -> Result<repo_proto::ListRepositoryMergeRequestsResponse, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/ListRepositoryMergeRequests",
            req,
        )
        .await
    }

    pub async fn register_repository_webhook_connect(
        &self,
        req: &repo_proto::RegisterRepositoryWebhookRequest,
    ) -> Result<repo_proto::RegisterRepositoryWebhookResponse, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/RegisterRepositoryWebhook",
            req,
        )
        .await
    }

    pub async fn delete_repository_webhook_connect(
        &self,
        req: &repo_proto::DeleteRepositoryWebhookRequest,
    ) -> Result<repo_proto::DeleteRepositoryWebhookResponse, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/DeleteRepositoryWebhook",
            req,
        )
        .await
    }

    pub async fn get_repository_webhook_status_connect(
        &self,
        req: &repo_proto::GetRepositoryWebhookStatusRequest,
    ) -> Result<repo_proto::WebhookStatus, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/GetRepositoryWebhookStatus",
            req,
        )
        .await
    }

    pub async fn get_repository_webhook_secret_connect(
        &self,
        req: &repo_proto::GetRepositoryWebhookSecretRequest,
    ) -> Result<repo_proto::WebhookSecret, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/GetRepositoryWebhookSecret",
            req,
        )
        .await
    }

    pub async fn mark_repository_webhook_configured_connect(
        &self,
        req: &repo_proto::MarkRepositoryWebhookConfiguredRequest,
    ) -> Result<repo_proto::MarkRepositoryWebhookConfiguredResponse, ApiError> {
        connect_call(
            self,
            "/proto.repository.v1.RepositoryService/MarkRepositoryWebhookConfigured",
            req,
        )
        .await
    }
}

